File: interview/run_gc_demo.py
import subprocess
import os
import matplotlib.pyplot as plt

# Define the Java program file and class name
JAVA_FILE = "GCDemo.java"
CLASS_NAME = "GCDemo"

# Define the list of GC options to test
GC_OPTIONS = [
    "-XX:+UseSerialGC",
    "-XX:+UseParallelGC",
    "-XX:+UseConcMarkSweepGC",
    "-XX:+UseG1GC",
    "-XX:+UseZGC",
    "-XX:+UseShenandoahGC",
    "-XX:+UseEpsilonGC"
]

# Function to compile the Java program
def compile_java_program():
    print("Compiling Java program...")
    result = subprocess.run(["javac", JAVA_FILE], capture_output=True, text=True)
    if result.returncode != 0:
        print("Compilation failed:")
        print(result.stderr)
        return False
    return True

# Function to run the Java program with the specified GC option
def run_gc_demo(gc_option):
    log_file = f"gc_demo_{gc_option.replace('+', '').replace(':', '').replace('XX', '').replace('Use', '').replace('GC', '').lower()}.log"
    print(f"Running GC demo with option {gc_option}...")
    with open(log_file, "w") as logfile:
        result = subprocess.run(["java", gc_option, CLASS_NAME], capture_output=True, text=True)
        logfile.write(result.stdout)
        logfile.write(result.stderr)
    return log_file

# Function to extract GC pause times from the log file
def extract_gc_times(log_file):
    gc_times = []
    with open(log_file, "r") as logfile:
        for line in logfile:
            if "Pause" in line:
                parts = line.split()
                time_ms = float(parts[-2])
                gc_times.append(time_ms)
    return sum(gc_times)

# Compile the Java program
if compile_java_program():
    gc_results = {}
    # Run the Java program with each GC option
    for gc_option in GC_OPTIONS:
        log_file = run_gc_demo(gc_option)
        total_gc_time = extract_gc_times(log_file)
        gc_results[gc_option] = total_gc_time

    # Plot the results
    plt.figure(figsize=(10, 6))
    plt.bar(gc_results.keys(), gc_results.values(), color='skyblue')
    plt.xlabel('GC Options')
    plt.ylabel('Total GC Pause Time (ms)')
    plt.title('Garbage Collection Performance Comparison')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    plt.savefig('gc_performance_comparison.png')
    plt.show()

else:
    print("Java program compilation failed. Exiting.")

File: interview/runGcDemo.kt
import java.io.File
import kotlin.system.exitProcess

// Define the Java program file and class name
const val JAVA_FILE = "GCDemo.java"
const val CLASS_NAME = "GCDemo"

// Define the list of GC options to test
val GC_OPTIONS = listOf(
    "-XX:+UseSerialGC",
    "-XX:+UseParallelGC",
    "-XX:+UseConcMarkSweepGC",
    "-XX:+UseG1GC",
    "-XX:+UseZGC",
    "-XX:+UseShenandoahGC",
    "-XX:+UseEpsilonGC"
)

// Function to compile the Java program
fun compileJavaProgram(): Boolean {
    println("Compiling Java program...")
    val process = ProcessBuilder("javac", JAVA_FILE).start()
    val exitCode = process.waitFor()
    if (exitCode != 0) {
        println("Compilation failed:")
        println(process.errorStream.bufferedReader().readText())
        return false
    }
    return true
}

// Function to run the Java program with the specified GC option
fun runGcDemo(gcOption: String): String {
    val logFile = "gc_demo_${gcOption.replace("+", "").replace(":", "").replace("XX", "").replace("Use", "").replace("GC", "").toLowerCase()}.log"
    println("Running GC demo with option $gcOption...")
    val process = ProcessBuilder("java", gcOption, CLASS_NAME).redirectOutput(File(logFile)).redirectError(File(logFile)).start()
    process.waitFor()
    return logFile
}

// Function to extract GC pause times from the log file
fun extractGcTimes(logFile: String): Double {
    val log = File(logFile)
    return log.readLines().filter { it.contains("Pause") }
        .sumOf { it.split(" ").last().removeSuffix("ms").toDouble() }
}

// Function to create plot using gnuplot
fun createPlot(gcResults: Map<String, Double>) {
    val plotData = "gc_results.dat"
    File(plotData).printWriter().use { out ->
        gcResults.forEach { (gcOption, time) ->
            out.println("${gcOption.replace("-XX:+", "")} $time")
        }
    }
    val plotScript = """
        set terminal png size 800,600
        set output 'gc_performance_comparison.png'
        set title 'Garbage Collection Performance Comparison'
        set xlabel 'GC Options'
        set ylabel 'Total GC Pause Time (ms)'
        set xtics rotate by -45
        plot '$plotData' using 2:xtic(1) with histograms title 'GC Pause Time' lc rgb 'skyblue'
    """.trimIndent()
    File("plot_script.gp").writeText(plotScript)
    ProcessBuilder("gnuplot", "plot_script.gp").start().waitFor()
}

fun main() {
    if (!compileJavaProgram()) {
        println("Java program compilation failed. Exiting.")
        exitProcess(1)
    }
    
    val gcResults = GC_OPTIONS.associateWith { gcOption ->
        val logFile = runGcDemo(gcOption)
        extractGcTimes(logFile)
    }

    createPlot(gcResults)
    println("GC performance comparison plot generated: gc_performance_comparison.png")
}
